jekyll/_cci2/language-python.md

```diff
@@ -2,104 +2,266 @@
+      - restore_cache:  # ensure this step occurs *before* installing dependencies
+          key: deps9-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
+      - run:
+        command: |
+          sudo pip install pipenv
+          pipenv install
+      - save_cache:
+          key: deps9-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
           paths:
-            - "venv"
-      - run: # run tests
-          command: |
-            . venv/bin/activate
-            python manage.py test
-      - store_artifacts: # special step to store test reports as artifacts
-          path: test-reports/
-          destination: tr1
-      # See https://circleci.com/docs/2.0/deployment-integrations/ for deploy examples    
+            - ".venv"
+            - "/usr/local/bin"
+            - "/usr/local/lib/python3.6/site-packages"
 ```
 
 {% endraw %}
 
-For the complete list of CircleCI configuration keys,
-refer to the [Configuring CircleCI]({{ site.baseurl }}/2.0/configuration-reference/) document.
+**Note:**
+Use the `chown` command
+to grant CircleCI access to dependency locations.
 
-## See Also
-{:.no_toc}
+### Run Tests
 
-- See the [Deploy]({{ site.baseurl }}/2.0/deployment-integrations/) document for example deploy target configurations.
-- See the [Caching Dependencies]({{ site.baseurl }}/2.0/caching/) document for caching strategies.
+Use the `run` step
+to run your test suite.
 
+```yaml
+version: 2
+jobs:
+  build:
+    # ...
+    steps:
+      # ...
+      - run:
+        command: |
+          pipenv run "python manage.py test"
+```
 
+### Upload And Store Test Results
 
+Use the [`store_test_results`]({{ site.baseurl }}/2.0/configuration-reference/#store_test_results) step
+to upload test results to CircleCI.
+These results will display in the **Test Summary** section of the CircleCI application.
+
+Use the [`store_artifacts`]({{ site.baseurl }}/2.0/configuration-reference/#store_artifacts) step
+to save test results as artifacts.
+
+```yaml
+version: 2
+jobs:
+  build:
+    # ...
+    steps:
+      # ...
+      - store_test_results:
+        path: test-results
+      - store_artifacts:
+        path: test-results
+        destination: tr1
+```
+
+### Deploy Application
+
+This Django application is not deployed anywhere.
+See the [Flask Project Walkthrough]({{ site.baseurl }}/2.0/project-walkthrough/) or the [Deploy]({{ site.baseurl }}/2.0/deployment-integrations/) document for deploy examples.
+
+## Full Configuration File
+
+{% raw %}
+
+```yaml
+version: 2
+jobs:
+  build:
+    working_directory: ~/circleci-demo-python-django
+    docker:
+      - image: circleci/python:3.6.4
+        environment:
+          PIPENV_VENV_IN_PROJECT: true
+          DATABASE_URL: postgresql://root@localhost/circle_test?sslmode=disable
+      - image: circleci/postgres:9.6.2
+        environment:
+          POSTGRES_USER: root
+          POSTGRES_DB: circle_test
+    steps:
+      - checkout
+      - run: sudo chown -R circleci:circleci /usr/local/bin
```

Comment:
Are these necessary, if so I think something is wrong in our images..

Reply:
oh, for caching.  We should be able to isolate pipenv's modules to a writable location, and thus avoiding the need to save/restore system-wide dependency locations..